src/SIMDMath/arithmetic.jl

```diff
@@ -0,0 +1,44 @@
+import Base: muladd
+
+@inline muladd(x::LVec{N, T}, y::LVec{N, T}, z::LVec{N, T}) where {N, T <: FloatTypes} = _muladd(x, y, z)
+@inline muladd(x::ScalarTypes, y::LVec{N, T}, z::LVec{N, T}) where {N, T <: FloatTypes} = muladd(constantvector(x, LVec{N, T}), y, z)
+@inline muladd(x::LVec{N, T}, y::ScalarTypes, z::LVec{N, T}) where {N, T <: FloatTypes} = muladd(x, constantvector(y, LVec{N, T}), z)
+@inline muladd(x::ScalarTypes, y::ScalarTypes, z::LVec{N, T}) where {N, T <: FloatTypes} = muladd(constantvector(x, LVec{N, T}), constantvector(y, LVec{N, T}), z)
+@inline muladd(x::LVec{N, T}, y::LVec{N, T}, z::ScalarTypes) where {N, T <: FloatTypes} = muladd(x, y, constantvector(z, LVec{N, T}))
+@inline muladd(x::ScalarTypes, y::LVec{N, T}, z::ScalarTypes) where {N, T <: FloatTypes} = muladd(constantvector(x, LVec{N, T}), y, constantvector(z, LVec{N, T}))
+@inline muladd(x::LVec{N, T}, y::ScalarTypes, z::ScalarTypes) where {N, T <: FloatTypes} = muladd(x, constantvector(y, LVec{N, T}), constantvector(z, LVec{N, T}))
+
+# muladd llvm instructions
+
+@inline function _muladd(x::LVec{2, Float64}, y::LVec{2, Float64}, z::LVec{2, Float64})
```

Comment:
can't you make a length generic version of this?

Reply:
Should be possible with a generated function though my quick attempt didn't work. Need to improve my meta-programming ability. Most of this is built up by pieces so there is some redundancy that I think I can simplify.

Reply:
Fully generic version...

```julia
const LLVMType = Dict{DataType, String}(
    Float32  => "float",
    Float64  => "double",
)

 @generated function _muladd(x::LVec{N, T}, y::LVec{N, T}, z::LVec{N, T}) where {N, T <: FloatTypes}
    s = """
        %4 = fmul contract <$N x $(LLVMType[T])> %0, %1
        %5 = fadd contract <$N x $(LLVMType[T])> %4, %2
        ret <$N x $(LLVMType[T])> %5
        """
    return :(
        llvmcall($s, LVec{N, T}, Tuple{LVec{N, T}, LVec{N, T}, LVec{N, T}}, x, y, z)
        )
end
```
```julia
julia> @code_llvm muladd(LVec{5, Float32}((1.1, 1.2, 1.2, 1.3, 1.65)), LVec{5, Float32}((1.1, 1.2, 1.5, 1.3, 1.5)), LVec{5, Float32}((1.1, 1.2, 1.1, 1.2, 1.5)))
;  @ /home/heltonmc/Documents/repos/Bessels_fork.jl/src/SIMDMath/arithmetic.jl:3 within `muladd`
define <5 x float> @julia_muladd_2505(<5 x float> %0, <5 x float> %1, <5 x float> %2) #0 {
top:
; ┌ @ /home/heltonmc/Documents/repos/Bessels_fork.jl/src/SIMDMath/arithmetic.jl:52 within `_muladd`
; │┌ @ /home/heltonmc/Documents/repos/Bessels_fork.jl/src/SIMDMath/arithmetic.jl within `macro expansion`
    %3 = fmul contract <5 x float> %0, %1
    %4 = fadd contract <5 x float> %3, %2
; └└
  ret <5 x float> %4

```

Reply:
add Float16 support?

Reply:
Seems reasonable to me for this. Done.
```julia
julia> @code_llvm muladd(LVec{5, Float16}((1.1, 1.2, 1.2, 1.3, 1.65)), LVec{5, Float16}((1.1, 1.2, 1.5, 1.3, 1.5)), LVec{5, Float16}((1.1, 1.2, 1.1, 1.2, 1.5)))
;  @ /home/heltonmc/Documents/repos/Bessels_fork.jl/src/SIMDMath/arithmetic.jl:3 within `muladd`
define <5 x half> @julia_muladd_435(<5 x half> %0, <5 x half> %1, <5 x half> %2) #0 {
top:
; ┌ @ none within `_muladd`
; │┌ @ none within `macro expansion`
    %3 = fmul contract <5 x half> %0, %1
    %4 = fadd contract <5 x half> %3, %2
; └└
  ret <5 x half> %4
```